runtime/error.c

```diff
@@ -1,12 +1,184 @@
+    char *errorMessage;
+    PC *pcs;
+} SimpleBacktrace;
+
+static void setBacktraceInternalError(void *data, const char *msg, int errnum);
+static void setBacktracePCInfoInternalError(void *data, const char *msg, int errnum);
+static int printBacktraceLine(void *data, PC pc, const char *filename, int lineno, const char *function);
+static int setPC(void *data, PC pc);
+static void getPCs(SimpleBacktrace *simpleBacktrace);
 
 TaggedPtr _bal_error_construct(TaggedPtr message, int64_t lineNumber) {
-    ErrorPtr ep = _bal_alloc(sizeof(struct Error));
+    SimpleBacktrace simpleBacktrace = {0, INITIAL_PC_COUNT, NO_ERROR};
+    void *p = malloc(sizeof(PC) * INITIAL_PC_COUNT);
+    if (p == NULL) {
+        simpleBacktrace.errorCode = NOT_ENOUGH_MEMORY_AVAILABLE;
+    }
+    else {
+        simpleBacktrace.pcs = p;
+        getPCs(&simpleBacktrace);
+    }
+
+    uint32_t nPCs = simpleBacktrace.nPCs;
+    uint64_t errorStructSize = sizeof(struct Error) + sizeof(PC) * nPCs;
+    ErrorPtr ep = _bal_alloc(errorStructSize);
     ep->message = message;
     ep->lineNumber = lineNumber;
+    ep->internalErrorCode = simpleBacktrace.errorCode;
+
+    if (p != NULL) {
+        ep->internalErrorMessage = simpleBacktrace.errorMessage;
+        ep->nPCs = nPCs;
+        PC *pcs = simpleBacktrace.pcs;
+        memcpy(ep->pcs, pcs, sizeof(PC) * nPCs);
+        free(pcs);
+    }
     return ptrAddFlags(ep, (uint64_t)TAG_ERROR << TAG_SHIFT);
 }
 
+static void getPCs(SimpleBacktrace *simpleBacktrace) {
+    if (state == NULL) {
+        state = backtrace_create_state(NULL, THREAD, setBacktraceInternalError, simpleBacktrace);
+        if (state == NULL) {
+            return;
+        }
+    }
+    backtrace_simple(state, 0, setPC, setBacktraceInternalError, simpleBacktrace);
+}
+
+// Implementation of backtrace_simple_callback
+static int setPC(void *data, PC pc) {
+    SimpleBacktrace *simpleBacktrace = data;
+    uint32_t nPCs = simpleBacktrace->nPCs;
+    uint32_t szPCs = simpleBacktrace->szPCs;
+    if (nPCs == szPCs) {
+        if (unlikely(szPCs == MAX_PC_COUNT)) {
+            simpleBacktrace->errorCode = EXCEEDS_MAX_PC_COUNT;
+            return FAIL;
+        }
+        szPCs = szPCs << 1;
+        simpleBacktrace->szPCs = szPCs;
+        void *p = realloc(simpleBacktrace->pcs, sizeof(PC) * szPCs);
+        if (p == NULL) {
+            simpleBacktrace->errorCode = NOT_ENOUGH_MEMORY_AVAILABLE;
+            return FAIL;
+        }
+        simpleBacktrace->pcs = p;
+    }
+
+    simpleBacktrace->pcs[nPCs] = pc;
+    simpleBacktrace->nPCs = nPCs + 1;
+    return SUCCESS;
+}
+
+// Implementation of backtrace_error_callback
+static void setBacktraceInternalError(void *data, const char *msg, int errnum) {
+    SimpleBacktrace *simpleBacktrace = data;
+    // the error that is created first is tracked
+    if (simpleBacktrace->errorCode != NO_ERROR) {
+        return;
+    }
+    simpleBacktrace->errorCode = errnum;
+    uint64_t messageLength = strlen(msg);
```

Comment:
There's a very bad error here (and in the other callback). You are not copying the terminating `\0`. This could cause invalid memory access when the message is accessed.

Furthermore this allocated memory is not being freed.